Add the option to store a custom Helm override file on the AMI's data volume

marcleblanc2 opened the pull request:
to enable customers to make customizations and have them persist on the data volume, instead of getting reset the next time they upgrade the AMI. For cases like [GRAPH-809](https://linear.app/sourcegraph/issue/GRAPH-809/scip-index-ingestion-failure-bad-connectionconn-closed).

jdpleiness reviewed (changes requested):
I'm getting a number of hits with `shellcheck` on the new changes that might need to be addressed (see below). Also can we make sure this has been tested on a few installs and upgrades and note that in the PR? 🙇 

```shell
➜  deploy git:(marc/add-AMI-custom-override-file-to-reboot-sh) ✗ shellcheck install/reboot.sh 

In install/reboot.sh line 13:
AMI_VERSION="$(cat $AMI_ROOT_VOLUME_VERSION_FILE)"
                   ^---------------------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
AMI_VERSION="$(cat "$AMI_ROOT_VOLUME_VERSION_FILE")"


In install/reboot.sh line 109:
$KUBECTL_CMD apply -f $DEPLOY_PATH/prometheus-override.ConfigMap.yaml
                      ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
$KUBECTL_CMD apply -f "$DEPLOY_PATH"/prometheus-override.ConfigMap.yaml


In install/reboot.sh line 114:
if [ -f $DEPLOY_PATH/sourcegraph-charts.tgz ]; then
        ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
if [ -f "$DEPLOY_PATH"/sourcegraph-charts.tgz ]; then


In install/reboot.sh line 115:
    $HELM_UPGRADE_INSTALL_CMD sourcegraph $DEPLOY_PATH/sourcegraph-charts.tgz
                                          ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
    $HELM_UPGRADE_INSTALL_CMD sourcegraph "$DEPLOY_PATH"/sourcegraph-charts.tgz


In install/reboot.sh line 122:
$KUBECTL_CMD create -f $DEPLOY_PATH/ingress.yaml
                       ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
$KUBECTL_CMD create -f "$DEPLOY_PATH"/ingress.yaml


In install/reboot.sh line 130:
if [ -f $DEPLOY_PATH/sourcegraph-executor-k8s-charts.tgz ]; then
        ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
if [ -f "$DEPLOY_PATH"/sourcegraph-executor-k8s-charts.tgz ]; then


In install/reboot.sh line 131:
    $HELM_UPGRADE_INSTALL_CMD executor $DEPLOY_PATH/sourcegraph-executor-k8s-charts.tgz
                                       ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
    $HELM_UPGRADE_INSTALL_CMD executor "$DEPLOY_PATH"/sourcegraph-executor-k8s-charts.tgz


In install/reboot.sh line 152:
    echo "$HELM_APP_VERSION" | sudo tee $AMI_ROOT_VOLUME_VERSION_FILE
                                        ^---------------------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
    echo "$HELM_APP_VERSION" | sudo tee "$AMI_ROOT_VOLUME_VERSION_FILE"


In install/reboot.sh line 160:
log "Script finishing after $(($(date +%s) - $START_TIME)) seconds"
                                             ^---------^ SC2004 (style): $/${} is unnecessary on arithmetic variables.

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
  https://www.shellcheck.net/wiki/SC2004 -- $/${} is unnecessary on arithmeti...
```

marcleblanc2 commented:
Made suggested changes from shellcheck linter.

Also:
- Moved the `helm repo update` command under a network connectivity test, because this command has an unconfigurable 2 minute timeout, which slows down startup for instances without internet connectivity
- Only run the `helm repo update` if the Helm charts are missing from the local disk
- Should we reorder these, and prioritize using the Helm charts from helm.sourcegraph.com if the instance can connect to it, and only use the local charts it the instance cannot connect to it?
- Tidied up the logic around getting version numbers from the root and data volumes, and comparing them
- Removed regex validation on version numbers, in case we want to use this with test builds
- Moved some repeated commands into functions

Tested this in each permutation and combination of:

1. Regular reboot (versions on data and root volumes match) vs upgrade (versions do not match)
2. With and without custom override file
3. With and without local Helm charts and connectivity to helm.sourcegraph.com

See the output of one of the test cases attached; I can attach the others as well. 
[8-upgrade-with-override-remote-charts.log](https://github.com/user-attachments/files/16689939/8-upgrade-with-override-remote-charts.log)

jdpleiness reviewed (approved):
Thanks for the follow-up 🚀 
Overall this is a huge improvement to the script! Thanks 🙏

marcleblanc2 commented:
Awesome, thank you! What're the odds this will make it into today's release?

Chickensoupwithrice commented:
> Awesome, thank you! What're the odds this will make it into today's release?

100% :)
